internal/file/watch/watch.go

```diff
@@ -0,0 +1,188 @@
+	"github.com/vdaas/vald/internal/log"
+	"github.com/vdaas/vald/internal/safety"
+)
+
+type Watcher interface {
+	Start(ctx context.Context) (<-chan error, error)
+	Add(dir string) error
+	Remove(dir string) error
+	Stop(ctx context.Context) error
+}
+
+type watch struct {
+	w        *fsnotify.Watcher
+	eg       errgroup.Group
+	dirs     []string
+	onChange func(ctx context.Context, name string) error
+	onCreate func(ctx context.Context, name string) error
+	onRename func(ctx context.Context, name string) error
+	onDelete func(ctx context.Context, name string) error
+	onWrite  func(ctx context.Context, name string) error
+	onChmod  func(ctx context.Context, name string) error
+	onError  func(ctx context.Context, err error) error
+}
+
+func New(opts ...Option) (Watcher, error) {
+	w := new(watch)
+	for _, opt := range append(defaultOpts, opts...) {
+		if err := opt(w); err != nil {
+			return nil, errors.ErrOptionFailed(err, reflect.ValueOf(opt))
+		}
+	}
+
+	if len(w.dirs) == 0 {
+		return nil, errors.ErrWatchDirNotFound
+	}
+	return w.init()
+}
+
+func (w *watch) init() (*watch, error) {
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		return nil, err
+	}
+	if w.w != nil {
+		err = w.w.Close()
+		if err != nil {
+			return nil, err
+		}
+	}
+	w.w = watcher
+
+	for _, dir := range w.dirs {
+		err = w.w.Add(dir)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return w, nil
+}
+
+func (w *watch) Start(ctx context.Context) (<-chan error, error) {
+	ech := make(chan error, 10)
+	w.eg.Go(safety.RecoverFunc(func() (err error) {
+		defer close(ech)
+		var (
+			event fsnotify.Event
+			ok    bool
+		)
+		handleErr := func(ctx context.Context, err error) {
+			log.Error(err)
+			select {
+			case <-ctx.Done():
+			case ech <- err:
+			}
+		}
+		for {
+			ok = true
+			err = nil
+			select {
+			case <-ctx.Done():
+				return ctx.Err()
+			case event, ok = <-w.w.Events:
+				if ok {
+					log.Debug("change detected file: ", event.Name)
+					if w.onChange != nil {
+						err = w.onChange(ctx, event.Name)
+						if err != nil {
+							handleErr(ctx, err)
+						}
+						err = nil
+					}
+					switch {
+					case event.Op&fsnotify.Write == fsnotify.Write && w.onWrite != nil:
+						log.Debug("Modified file: ", event.Name)
+						err = w.onWrite(ctx, event.Name)
+					case event.Op&fsnotify.Create == fsnotify.Create && w.onCreate != nil:
+						log.Debug("Created file: ", event.Name)
+						err = w.onCreate(ctx, event.Name)
+					case event.Op&fsnotify.Remove == fsnotify.Remove && w.onDelete != nil:
+						log.Debug("Removed file: ", event.Name)
+						err = w.onDelete(ctx, event.Name)
+					case event.Op&fsnotify.Rename == fsnotify.Rename && w.onRename != nil:
+						log.Debug("Renamed file: ", event.Name)
+						err = w.onRename(ctx, event.Name)
+					case event.Op&fsnotify.Chmod == fsnotify.Chmod && w.onChmod != nil:
+						log.Debug("File changed permission: ", event.Name)
+						err = w.onChmod(ctx, event.Name)
+					}
+				}
+			}
```

Comment:
![badge](https://img.shields.io/badge/review-suggestion-blue.svg?style=flat-square)
```suggestion
			case err, ok = <-w.w.Errors:
			}
```
and remove the goroutine started from L145.
it's better to reduce the number of useless goroutines.

Reply:
fsnotify recommend to watch event and error in a different goroutine.
https://github.com/howeyc/fsnotify/issues/7

Reply:
I read it but i couldn't found a recommendation about that.
i think it's just about "not to create a fsnotify Watcher and monitor channels in a same goroutine".

I'm considering about the race condition about `init()` func when the `w.w.Errors` and `w.w.Events` closed simultaneously.

Reply:
sorry they recommend this thing in README.md
```
Do I have to watch the Error and Event channels in a separate goroutine?

As of now, yes. Looking into making this single-thread friendly (see howeyc #7)
```
and I agree with init race condition, I'll fix race condition problem about it but I think it's okay to run in different goroutine that mentioned in howeyc/fsnotify#7

Reply:
as described in README.md of fsnotify, it is only needed "*a* separate goroutine".
we don't have to have two goroutines here.

Reply:
a separate goroutine doesn't mean using different goroutine?